File: frontend/src/api/exam.ts
import { apiGet, apiPost, apiPostForm } from './client';

export interface ExamTypeMeta {
  key: string;
  label: string;
}

export interface ExamMeta {
  exam_types: ExamTypeMeta[];
  sections: ExamTypeMeta[];
  mock_structure: { section: string; count: number }[];
}

export interface ExamQuestionItem {
  id: string;
  exam_type: string;
  section: string;
  question: string;
  options: Record<string, string>;
  audio_text: string;
  difficulty: string;
  source: string;
  answer?: string;
  analysis?: string;
}

export interface PracticeCheckResult {
  ok: boolean;
  correct: boolean;
  answer: string;
  analysis: string;
  question: ExamQuestionItem;
  mistake_archived?: boolean;
}

export interface MockPaper {
  run_id: string;
  paper_id: string;
  title: string;
  duration_minutes: number;
  structure: { section: string; question_ids: string[] }[];
  questions: ExamQuestionItem[];
}

export interface MockResult {
  ok: boolean;
  run_id: string;
  score: number;
  section_scores: Record<string, { total: number; correct: number; score: number }>;
  detail: {
    question_id: string;
    section: string;
    correct: boolean;
    my_answer: string;
    answer: string;
    analysis: string;
    ratio?: number;
  }[];
  mistakes_archived: number;
}

export interface MockHistoryItem {
  run_id: string;
  exam_type: string;
  score: number;
  section_scores: Record<string, { total: number; correct: number }>;
  finished_at: string;
}

export interface ExamWord {
  id: string;
  word: string;
  phonetic: string;
  meaning: string;
  example: string;
  freq_rank: number;
}

export interface EssayGradeResult {
  ok: boolean;
  score: number;
  dimensions: { name: string; score: number; comment: string }[];
  sentence_feedback: { original: string; revised: string; reason: string }[];
  highlights: string[];
  suggestions: string[];
}

export interface ListeningMaterial {
  ok: boolean;
  title: string;
  transcript: string;
  sentences: string[];
  blanks: { sentence_index: number; word: string }[];
  translation: string;
}

export interface ChallengeStatus {
  active: boolean;
  id?: string;
  name?: string;
  exam_type?: string;
  days_total?: number;
  days_done?: number;
  checkins?: string[];
  checked_today?: boolean;
  today_progress?: number;
  today_goal?: number;
  can_checkin?: boolean;
}

const AI_TIMEOUT = { timeoutMs: 180_000 };

export const fetchExamMeta = () => apiGet<ExamMeta>('/api/exam/meta');

export const fetchBankSummary = (exam_type: string) =>
  apiGet<{ exam_type: string; sections: Record<string, number> }>(
    `/api/exam/bank/summary?exam_type=${encodeURIComponent(exam_type)}`,
  );

export const generateQuestions = (exam_type: string, section: string, count = 5) =>
  apiPost<{ ok: boolean; created: number }>('/api/exam/generate', { exam_type, section, count }, AI_TIMEOUT);

export const importQuestions = (exam_type: string, file: File) => {
  const form = new FormData();
  form.append('exam_type', exam_type);
  form.append('file', file);
  return apiPostForm<{ ok: boolean; imported: number }>('/api/exam/import', form, AI_TIMEOUT);
};

export const fetchPracticeQuestions = (exam_type: string, section: string, count = 5) =>
  apiGet<{ questions: ExamQuestionItem[] }>(
    `/api/exam/practice?exam_type=${encodeURIComponent(exam_type)}&section=${encodeURIComponent(section)}&count=${count}`,
    AI_TIMEOUT,
  );

export const checkPracticeAnswer = (question_id: string, answer: string, archive_wrong = true) =>
  apiPost<PracticeCheckResult>('/api/exam/practice/check', { question_id, answer, archive_wrong });

export const logPractice = (payload: {
  exam_type?: string;
  section?: string;
  activity?: string;
  total?: number;
  correct?: number;
  meta?: Record<string, unknown>;
}) => apiPost<{ ok: boolean }>('/api/exam/practice/log', payload);

export const startMock = (exam_type: string) =>
  apiPost<MockPaper>('/api/exam/mock/start', { exam_type }, AI_TIMEOUT);

export const submitMock = (run_id: string, answers: Record<string, string>) =>
  apiPost<MockResult>('/api/exam/mock/submit', { run_id, answers }, AI_TIMEOUT);

export const fetchMockHistory = () => apiGet<MockHistoryItem[]>('/api/exam/mock/history');

export const fetchWords = (exam_type: string, offset = 0, limit = 20) =>
  apiGet<{ total: number; words: ExamWord[] }>(
    `/api/exam/words?exam_type=${encodeURIComponent(exam_type)}&offset=${offset}&limit=${limit}`,
  );

export const seedWords = (exam_type: string, count = 30) =>
  apiPost<{ ok: boolean; added: number }>('/api/exam/words/seed', { exam_type, count }, AI_TIMEOUT);

export const collectWord = (word_id: string) =>
  apiPost<{ ok: boolean; card_id: string }>(`/api/exam/words/${word_id}/collect`, {});

export const gradeEssay = (payload: { exam_type: string; kind: 'writing' | 'translation'; prompt?: string; text: string }) =>
  apiPost<EssayGradeResult>('/api/exam/essay/grade', payload, AI_TIMEOUT);

export const fetchListeningMaterial = (exam_type: string, topic = '') =>
  apiPost<ListeningMaterial>('/api/exam/listening/material', { exam_type, topic }, AI_TIMEOUT);

export const fetchChallengeStatus = () => apiGet<ChallengeStatus>('/api/exam/challenge');

export const joinChallenge = (exam_type = 'cet4') =>
  apiPost<{ ok: boolean; id: string }>('/api/exam/challenge/join', { exam_type });

export const challengeCheckin = () =>
  apiPost<{ ok: boolean; already?: boolean; days_done: number; finished?: boolean; points_earned?: number; points?: number }>(
    '/api/exam/challenge/checkin',
    {},
  );

/** 讯飞 TTS：返回可直接播放的 data URL */
export async function ttsToDataUrl(text: string): Promise<string> {
  const res = await apiPost<{ mime: string; audio_base64: string }>('/api/tts', { text }, { timeoutMs: 60_000 });
  return `data:${res.mime};base64,${res.audio_base64}`;
}

File: backend/app/models/exam.py
"""考级中心：题库 / 试卷 / 模考 / 词书 / 打卡挑战。"""
from datetime import datetime
from uuid import uuid4

from sqlalchemy import DateTime, Float, Integer, JSON, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.session import Base

EXAM_TYPES = ("cet4", "cet6", "ielts", "cantonese")
EXAM_SECTIONS = ("listening", "reading", "cloze", "translation", "writing", "vocab")


class ExamQuestion(Base):
    __tablename__ = "exam_questions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    exam_type: Mapped[str] = mapped_column(String(24), index=True, nullable=False)
    section: Mapped[str] = mapped_column(String(24), index=True, nullable=False)
    question: Mapped[str] = mapped_column(Text, nullable=False, default="")
    # 客观题选项 {"A": "...", "B": "..."}；主观题为空
    options: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    answer: Mapped[str] = mapped_column(Text, nullable=False, default="")
    analysis: Mapped[str] = mapped_column(Text, nullable=False, default="")
    # 听力题的朗读原文（TTS 合成用）
    audio_text: Mapped[str] = mapped_column(Text, nullable=False, default="")
    difficulty: Mapped[str] = mapped_column(String(16), nullable=False, default="medium")
    source: Mapped[str] = mapped_column(String(16), nullable=False, default="ai")  # ai | import
    created_by: Mapped[str] = mapped_column(String(36), nullable=False, default="")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class ExamPaper(Base):
    __tablename__ = "exam_papers"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    exam_type: Mapped[str] = mapped_column(String(24), index=True, nullable=False)
    title: Mapped[str] = mapped_column(String(256), nullable=False, default="")
    # [{"section": "listening", "question_ids": [...]}, ...]
    structure: Mapped[list] = mapped_column(JSON, nullable=False, default=list)
    duration_minutes: Mapped[int] = mapped_column(Integer, nullable=False, default=60)
    source: Mapped[str] = mapped_column(String(16), nullable=False, default="ai")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class ExamMockRun(Base):
    __tablename__ = "exam_mock_runs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), index=True, nullable=False)
    paper_id: Mapped[str] = mapped_column(String(36), index=True, nullable=False)
    exam_type: Mapped[str] = mapped_column(String(24), nullable=False, default="")
    answers: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    # {"listening": {"total": 10, "correct": 7}, ...}
    section_scores: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="ongoing")  # ongoing | done
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class ExamPracticeLog(Base):
    """专项刷题 / 打字 / 听写等训练记录（供周报与挑战统计）。"""

    __tablename__ = "exam_practice_logs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), index=True, nullable=False)
    exam_type: Mapped[str] = mapped_column(String(24), nullable=False, default="")
    section: Mapped[str] = mapped_column(String(24), nullable=False, default="")
    activity: Mapped[str] = mapped_column(String(24), nullable=False, default="practice")  # practice | typing | dictation | essay
    total: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    correct: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    meta_json: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class ExamWordEntry(Base):
    __tablename__ = "exam_word_entries"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    exam_type: Mapped[str] = mapped_column(String(24), index=True, nullable=False)
    word: Mapped[str] = mapped_column(String(128), index=True, nullable=False)
    phonetic: Mapped[str] = mapped_column(String(128), nullable=False, default="")
    meaning: Mapped[str] = mapped_column(Text, nullable=False, default="")
    example: Mapped[str] = mapped_column(Text, nullable=False, default="")
    freq_rank: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class ChallengeCampaignRecord(Base):
    """21 天打卡挑战：每用户一条参与记录。"""

    __tablename__ = "challenge_campaigns"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), index=True, nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False, default="21 天备考挑战")
    exam_type: Mapped[str] = mapped_column(String(24), nullable=False, default="cet4")
    days_total: Mapped[int] = mapped_column(Integer, nullable=False, default=21)
    # 每日目标：{"words": 10, "questions": 5}
    daily_goal: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    # ["2026-08-13", ...]
    checkins: Mapped[list] = mapped_column(JSON, nullable=False, default=list)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="active")  # active | done | failed
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
